enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+6. The "new" operator "adopts" the CSI driver objects. Namely, it updates DaemonSet of the CSI driver, which then does
+   rolling update of the driver pods, using supported images.
+
+To adopt the old objects, the new operator must use the same names for all objects created by the old operator.
+
+### [efs-utils](https://github.com/aws/efs-utils)
+
+We need to get efs-utils into the container with the EFS CSI driver to enable data encryption in transport. efs-utils
+are just two Python scripts, which create / monitor stunnel connections between OCP nodes and NFS server somewhere in
+AWS cloud.
+
+We've chosen to use base image approach. There are two reasons: 1) we know the drill, and 2) we don't want
+to support a generic RPM that can be used outside of OCP.
+
+1. Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+2. Create a new base image with the utilities, say `ose-aws-efs-utils-base`:
+   ```Dockerfile
+   FROM: ocp4.8:base
+   RUN yum install stunnel python <and any other deps>
+   COPY <the utilities> /usr/bin
+   ```
+3. The EFS CSI driver then uses it as the base image:
+   ```Dockerfile
+   FROM registry.ci.openshift.org/openshift/release:golang-1.15 AS builder
+   ... build the image ...
+
+   FROM ose-aws-efs-utils-base
+   COPY --from=builder <the driver binary> /usr/bin/
+   ```
+
+It would be nice if we did not ship `ose-aws-efs-utils-base` anywhere, so people can consume the EFS utils only through
+the EFS operator. **We do not want to maintain it for everyone.**
+
+### Code organization
+
+* Reuse existing `github.com/openshift/aws-efs-operator` for the operator.
+  * Move the controller-runtime based operator one to a separate branch (`release-4.7`, even though it contains all
+    previous releases?)
+  * Continue with library-go based one in `master`.
+* Fork the CSI driver into `github.com/openshift/aws-efs-csi-driver`.
+* Fork `github.com/aws/efs-utils` into `github.com/openshift/aws-efs-utils`.
+
+### Open Questions
+
+* How to graduate community operator into a supported one?
+  * The community one does not go through registry.redhat.io.
+    * The "community" team is willing to do anything that's necessary to help with the transition.
+  * The community one has only a single channel, `stable`.
+  * We want the community one working in 4.7 and older and the supported one in 4.8 and newer.
+  * It should be possible to ship new community updates in 4.7 channel/branch to fix potential bugs there and/or
+    make the transition as seamless as possible.
+
+### Test Plan
+
+* Automatic jobs:
+  * Run CSI certification tests with the CSI driver:
+     1. Install OCP on AWS.
+     2. Install the operator.
+     3. Create AWS EFS share (this may require extra privileges in CI and figure out the networking!)
+     4. Set up dynamic provisioning in the driver (by creating a StorageClass). This assumes dynamic provisioning
+        is implemented (released) upstream before we start testing.
```

Comment:
I recommend not tying the test plan to dynamic provisioning -- or in fact any direct AWS interaction -- in the actual driver or operator. The test environment itself can be configured with appropriate access to the AWS side; and the test cases can do the provisioning/deprovisioning as separate steps from the operator intsallation and configuration.

We had to do plenty of this while developing the community operator. We never quite got to automating tests end to end, but I wrote a [utility](https://github.com/2uasimojo/aws-efs-manage) that does the AWS networking, security, and provisioning setup/teardown. You may be able to reuse pieces of it here.

As for permissions, I'm almost certain it is already possible to specify AWS creds in CI today. That's all you should need.